### ISSUE TITLE
Developer guidelines

### ISSUE THREAD

**Issue by langner** [cclib/cclib#555]
It'd be useful to have some developer guidelines in our docs, so that we can point contributors to it, and also achieve consensus among ourselves and make our lives easier. This would include at least:
1. Git workflow policies, including merging/rebasing conventions or rules
1. Review workflow, including our unwritten rule that you can't merge your own code (with whatever exceptions we deem appropriate)
1. Style recommendations

**Comment by langner** [comment]
4. Procedures for unit test files and code.
5. Procedures for regression tests and files in cclib/cclib-data

**Comment by langner** [comment]
6. Parsing conventions, such as not predicating parsing on previously parsed attributes. So, parse data multiple times if available, even if we don't use the value for attributes. OTOH, prefer overwriting re-parsed attributes, or at least checking their consistency, via `set_attribute`.

**Comment by berquist** [comment]
Rebasing formatting changes is making it too difficult to review inside a PR. If we are unhappy with the number of commits in a PR, we should either:
1. "Squash and merge"
2. Request an interactive rebase to squash those commits just before merging the PR.

**Comment by langner** [comment]
7. Describe our approach to parsing files, streams and multiples files and/or directories.

**Comment by langner** [comment]
Not for points 4 & 5: when updating unit test files, first copy the newest version of the test logfiles to a new directory and then update them with the actual new contents. That way the history in git shows the changes from the last version by default, which makes things sometimes easier to debug. This was done recently when updating ORCA in #575 and works nicely.

**Comment by langner** [comment]
8. Describe our philosophy for methods: willing to include anything that relies on attributes and physical constants. Methods should also be pure Python and reasonable light.

**Comment by berquist** [comment]
From https://github.com/cclib/cclib/pull/568#discussion_r212835180,

> let's have a convention to use only one tool if we want to auto-format. We have enough pep8 violations that it would make sense to format in bulk once if we want to go in that direction.

**Comment by berquist** [comment]
Now that docs are easy to modify again, how would you like to do this? Will everything go in the already existing `development.rst`?

**Comment by langner** [comment]
> Now that docs are easy to modify again, how would you like to do this? Will everything go in the already existing development.rst?

I dunno, haven't thought about the details yet. One thought: best to do it incrementally, so it's easier to review and have a discussion (where needed).

Also...

9. From cclib/cclib.github.io#21: I think we need to outline in the dev section what we expect from pull requests and point people to that when they send very large changes, etc.

**Comment by langner** [comment]
10. As mentioned in cclib/cclib.github.io#15, provide some guidance on how to go about adding new parsers.

**Comment by langner** [comment]
11. Explain the mechanics of making new releases.

**Comment by langner** [comment]
12. Explain what the goals and non-goals are, in particular that we're not concerned with correctness of data. The goal is to consistently parse data, and not vouch for it.

**Comment by langner** [comment]
13. Add expectations about using the logger as per #237.

**Comment by langner** [comment]
14. Stick to NumPy style docstrings, or some other format we want, as mentioned in #694.

**Comment by berquist** [comment]
15. When skipping lines, prefer using `skip_lines` and `skip_line` with a short message about line contents over repeated calls to `next(inputfile)`.

**Comment by berquist** [comment]
As part of style (3) and PR size/expectations (9), unrelated changes should in separate PRs, including non-functional changes. Style changes aren't discouraged but they need a separate discussion. Smaller PRs make it easier to do this and for them to be reviewed and merged in general.

**Comment by berquist** [comment]
17.  When updating a PR branch with the latest changes from `master` or `main`, prefer to rebase rather than merge.
18.  When receiving PR reviews, rebasing makes conversations hard to follow since they are no longer properly associated with their parent commit (this is a GitHub, not Git, problem).  Try to hold off on rebasing against the main branch as long as possible.

**Comment by berquist** [comment]
19. When adding test data for nested programs, where the desired output appears inside the output of another, the target program is the outer folder and the calling program is the inner folder.  See https://github.com/cclib/cclib/pull/1230 for an example, where the layout is `data/NBO/basicNBO7.0/basicORCA5.0/` for NBO run by ORCA.

**Comment by berquist** [comment]
20. How to perform development (edit code, run tests) using Dev Containers (see #1212, https://github.com/cclib/cclib/pull/1251#issuecomment-1685370562).

**Comment by shivupa** [comment]
sorry autoclosed by PR merge
